java/src/main/java/ai/rapids/cudf/HostColumnVector.java

```diff
@@ -929,74 +929,127 @@ public ColumnBuilder appendStructValues(StructData... inputList) {
+     */
+    private void growStringBuffersAndRows(int stringLength) {
+      assert rows + 2 <= Integer.MAX_VALUE : "Row count cannot go over Integer.MAX_VALUE";
+      rows++;
+
+      if (offsets == null) {
+        // Initialize data buffer with at least 64 bytes to avoid growing too frequently.
+        data = HostMemoryBuffer.allocate(Math.max(64, stringLength));
+        offsets = HostMemoryBuffer.allocate((estimatedRows + 1) * OFFSET_SIZE);
+        offsets.setInt(0, 0);
+        rowCapacity = estimatedRows;
+        return;
       }
-      if (hasNull || nullCount > 0) {
-        if (valid == null) {
-          long targetValidSize = ColumnView.getNativeValidPointerSize((int)estimatedRows);
-          valid = HostMemoryBuffer.allocate(targetValidSize);
-          valid.setMemory(0, targetValidSize, (byte) 0xFF);
-        } else if (valid.length < ColumnView.getNativeValidPointerSize((int)rows)) {
-          long newValidLen = valid.length * 2;
-          HostMemoryBuffer newValid = HostMemoryBuffer.allocate(newValidLen);
-          newValid.setMemory(0, newValidLen, (byte) 0xFF);
-          valid = copyBuffer(newValid, valid);
-        }
+
+      if (rows > rowCapacity) {
+        long newCap = Math.min(rowCapacity * 2, Integer.MAX_VALUE - 2);
+        offsets = copyBuffer(HostMemoryBuffer.allocate((newCap + 1) * OFFSET_SIZE), offsets);
+        rowCapacity = newCap;
+      }
+
+      long currentLength = currentByteIndex + stringLength;
+      long requiredLength = data.length;
+      while (currentLength > requiredLength) {
+        requiredLength = requiredLength * 2;
+      }
+      if (requiredLength > data.length) {
+        data = copyBuffer(HostMemoryBuffer.allocate(requiredLength), data);
+      }
+    }
+
+    /**
+     * For struct columns, we only need to update rows and rowCapacity (for the growth of
+     * valid buffer), because struct columns hold no buffer itself.
+     * Please call this method before appending any value or null.
+     */
+    private void growStructBuffersAndRows() {
+      assert rows + 1 <= Integer.MAX_VALUE : "Row count cannot go over Integer.MAX_VALUE";
+      rows++;
+
+      if (rowCapacity == 0) {
+        rowCapacity = estimatedRows;
+      } else if (rows > rowCapacity) {
+        rowCapacity = Math.min(rowCapacity * 2, Integer.MAX_VALUE - 1);
       }
     }
 
+    /**
+     * The Java substitution of native method `ColumnView.getNativeValidPointerSize`.
+     * Ideally, this method can speed up growValidBuffer by eliminating the JNI call.
+     */
+    private static long byteSizeOfNullMask(int numRows) {
+      // number of bytes required = Math.ceil(number of bits / 8)
+      int actualBytes = (numRows >> 3) + Math.min(numRows % 8, 1);
+
+      // padding to multiply of 64 bytes, just as cuDF default padding boundary
+      long padding = actualBytes >> 6;
+      if (actualBytes % 64 > 0) padding++;
+      padding = padding << 6;
```

Comment:
Moving this to java is a really good win for performance, but I think we can still do better. The C++ code is.

```c++
  CUDF_EXPECTS(padding_boundary > 0, "Invalid padding boundary");
  auto necessary_bytes = cudf::util::div_rounding_up_safe<size_type>(number_of_bits, CHAR_BIT);

  auto padded_bytes = padding_boundary * cudf::util::div_rounding_up_safe<size_type>(
                                           necessary_bytes, padding_boundary);
  return padded_bytes;
```
We could try to copy what cudf does for `div_rounding_up_safe` but, we know for this particular case that 
`CHAR_BIT` is 8,  `padding_boundary` is 64 and `size_type` is `int32_t`. Which means we can use a `long`
to calculate the intermediate results to get a `div_rounding_up_safe` that is much faster than needing to use
conditions.

```java
long necessaryBytes = ((long)numberOfBits + numberOfBits - 1)/8;
long paddedBytes = (necessaryBytes + necessaryBytes - 1)/64;
return paddedBytes;
```

Reply:
Great suggestion. I've applied it.